### ISSUE TITLE
[ban-types] Default options are never removed

### ISSUE THREAD

**Issue by OliverJAsh** [typescript-eslint/typescript-eslint#686]
**Repro**

```js
module.exports = {
    parser: '@typescript-eslint/parser',
    plugins: ['@typescript-eslint'],
    rules: {
        '@typescript-eslint/ban-types': [
            'error',
            {
                types: {
                    Omit: 'Prefer `OmitStrict`.',
                },
            },
        ],
    },
};
```

```TS
// Don't use 'Object' as a type. Use Record<string, any> instead
type Foo1 = Object;
// Don't use 'Boolean' as a type. Use boolean instead
type Foo2 = Boolean;
// Don't use 'Number' as a type. Use number instead
type Foo3 = Number;
```

**Expected Result**
No errors, because I am overriding the [default options](https://github.com/typescript-eslint/typescript-eslint/blob/e325b728098f73b98ada8ec05ea0175403c80d1f/packages/eslint-plugin/src/rules/ban-types.ts#L83).

**Actual Result**
Each line errors. It is not possible to disable the default options.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.11.0` |
| `@typescript-eslint/parser`        | `1.11.0` |
| `TypeScript`                       | `3.5.2` |
| `ESLint`                           | `6.0.1` |
| `node`                             | `8.15.1` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
Someone brought this up a few days ago on an unrelated thread! You've run into a bug that occurs because of the default options merging.

Out of curiosity - why don't you want the default options? Why do you want to use the capital types instead of the lower-case ones?

**Comment by OliverJAsh** [comment]
No good reason. Just didn't expect to have them forced on me. They're only supposed to be defaults after all.

**Comment by adi-zz** [comment]
I agree it would be nice to be able to ban **only** your selected types and get rid of the defaults, but it's also useful to be able to add some types to be banned and still have the defaults banned (without having to add [them](https://github.com/typescript-eslint/typescript-eslint/blob/4edf0d71ef0a8a3370165e804dd693033d6207bc/packages/eslint-plugin/src/rules/ban-types.ts#L64-L86) again in the rule configuration). So maybe config like this could be supported:

        '@typescript-eslint/ban-types': [
            'error',
            {
                types: {
                    Omit: 'Prefer `OmitStrict`.',
                },
                keepDefaults: true, // keep the default settings
            },
        ]

**Comment by mysticatea** [comment]
Looks good if the default behavior is separated into another rule that has a more self-described name. For example, `no-wrapper-object-types`, `prefer-primitive-types`, or something like. In core, [no-restricted-globals](https://eslint.org/docs/rules/no-restricted-globals)-like rules don't have defaults.

**Comment by cedx** [comment]
In Angular projects, in order to support SSR, we constantly use the `PLATFORM_ID` injection token, which is typed as `Object`:
https://angular.io/api/core/PLATFORM_ID

In the source code, its value is a `string`, so using `Record<string, any>` instead of `Object` is not an option (the TS compiler complains because strings are not records, nor `object`).

It would be useful to disable some default options of this rule.

**Comment by joebowbeer** [comment]
I'm OK with the default rule, e.g., for Object, but I don't like that there's no way to disable (or change) the `fixWith` that replaces them all with `Record<string, any>`.

**Comment by threehams** [comment]
I have about 30 people migrating from Flow, where `Object` means `any object` (or really, means `any` as of version 80 or so). So, I don't want them to use an equally weak `Record<string, any>` - I want them to write a meaningful type, like `{ name: string, age: number }`.

For anyone else hitting this, I'm getting around the default fixer by setting `fixWith: "Object"`.

**Comment by bradzacher** [comment]
# YOU DO NOT WANT TO DO THAT

`Object` is almost the same as `any` in typescript.

```ts
function foo(a: Object) { }

foo('');
foo(1);
foo(false);
foo({});
foo(null); // error: Argument of type 'null' is not assignable to parameter of type 'Object'.
foo(undefined); // error: Argument of type 'undefined' is not assignable to parameter of type 'Object'.
```
https://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABMOcAUBDAXIg8gIwCsBTaASkQG9EBfAKDpXQHJmyBuR1NARg6-TAMAGwDOxfkzSUak7mBDDhc9OAAmxYDDDE1-IA

`Record<string, unknown>` is a good alternative that means any object, but forces the consumer to check the values.

**Comment by threehams** [comment]
Sorry, miscommunication! I'm using `fixWith: Object`, but keeping it an error, to stop it from being automatically fixed to `Record`. It's just a way of getting around the built-in fixer.

**Comment by bradzacher** [comment]
The bug is in the default options code, as the code wasn't designed with a two-level object config in mind (there wasn't any config like that at the time).

The code just assigns any missing properties from the default options to the user options.
So setting `fixWith: undefined` means the key exists, and won't be overridden.

**Comment by threehams** [comment]
Mind if I work on this?

**Comment by bradzacher** [comment]
I suspect that it will be near impossible to fix in a generic way.
I think this rule will just have to be a special case that doesn't use the options merging infra.

The problem is essentially this: in a nutshell, the default options merging function does `Object.assign(userOptions ?? {}, defaultOptions)`. This works great for 99.99% of rules, which have options objects that are simple, and consistent in how they should be merged (regardless of user input).

`ban-types` is different - it should be treated as an "all or nothing" merge (i.e. `userOptions ?? Object.assign({}, defaultOptions)`).

Happy to accept a PR, but FYI this will be a breaking change, so it'll have to wait to be merged. We haven't planned when we wanted to do the 3.0.0 release yet.

**Comment by olee** [comment]
How about allowing users to add a configuration like this?`
```
                '@typescript-eslint/ban-types': ['warn', {
                    types: {
                        Object: false,
                    }
                }],
```

So by explicitly setting an option to false, it could be disabled.
This could easily be implemented by adding a check like `if (bannedType === false) continue;` below this line https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/ban-types.ts#L127

**Comment by HolgerJeromin** [comment]
BTW we need this change because we have an own type `HMI.Symbol` which is used often inside the HMI namespace without the namespace. 
Perhaps we should, but right now I can not activate this rule (and overlooked some `let num: Number` things :)

**Comment by bradzacher** [comment]
FYI - just merged a PR which adds the `extendDefaults: boolean` option.
You can set this to `false` if you want to define your own ban list without including the default options.
